### ISSUE TITLE
Error in correlations generating a dataset with gen_output.generate()

### ISSUE THREAD

**Issue by soyunalmapresente** [CALMDataGenerator/Calm-Data-Generator#12]
**Describe the bug**
Error in correlations when generating a dataset with gen_output.generate(). The input correlations are very different from the correlations on data generated.
I updated Calm-data-generator but it seems version number has not been changed. 
VERSIONS. python:  3.11.9 numpy:  1.26.4 scipy:  1.14.1 matplotlib:  3.10.0 Image:  12.0.0 HDF5 file format:  3.15.1
iPython:  9.10.0  jupiter notebook:  7.5.3  jupiter lab:  4.5.4  scikit-learn:  1.7.2  pytorch:  2.2.2  tabpfn:  6.4.1
pydot:  4.0.1  graphviz:  0.21  torchviz:  0.0.3  skorch:  1.3.1  seaborn:  0.13.2  shap:  0.49.1  joblib:  1.5.3
calm_data_generator:  2.2.1

**To Reproduce**  
----------------------------------------------------------------
``````
import scipy.stats as stats
from calm_data_generator.generators.clinical.Clinic import ClinicalDataGenerator
from calm_data_generator.tutorials.clinical_generator import build_correlation_matrix

gen_output = ClinicalDataGenerator()

n_demo = 4
custom_demo = {
    "INR": {"distribution": "truncnorm", "a": 0, "b": 5, "loc": 1.1, "scale": 0.3},
    "batch": {"distribution": "randint", "low": 1, "high": 5}, # 1,2,3,4
}
group_sizes = [100, 200, 300, 400]

correlations_config = [
    {"internal": (0.3, 0.6), "demo_idx": 0, "demo_corr": 0.4},   # A & A <-> age
    {"internal": (0.2, 0.4), "demo_idx": 1, "demo_corr": 0.4},   # B & B <-> sex
    {"internal": (0.2, 0.3)},                                    # D
    {"internal": 0.0},                                           # NS
]

corr_mx = build_correlation_matrix(n_demo, group_sizes, correlations_config)

datasets = gen_output.generate(
    n_samples=N_SAMPLES, # n_patients does not work
    n_genes=sum(group_sizes),
    custom_demographic_columns=custom_demo,
    demographic_gene_correlations=corr_mx,

    target_variable_config={ "weights": {
            "Age":         0.3 * 10,   # r ≈ 0.3
            "Sex_Binario": 0.1 * 10,   # r ≈ 0.1
            "G_0":         0.2 * 10,   # r ≈ 0.2  (representante grupo A)
            "G_100":       0.5 * 10},   # r ≈ 0.5  (representante grupo B)
        },
    )
``````

**Expected behavior**
Keep correlations defined at generation. I just changed the name of the variables.

Arrange unique matrix MCY_df:
MCY_df = genes_df.join( demog_df )
RENAME G_0:G_99  -----> A_gene1:A_gene100,  
       G_100:G_299  ------> B_gene101:B_gene300,  
       G_300:G_599  ------> D_gene301:D_gene600,
       G_600:G_1000  ------> NS_gene601:NS_gene_1000
       Age 	    ------> age
       Sex        ------> sex
       INR 	     ------>  INR
       batch     ------>  batch
       Group    ------> Y (disease)
var_names = generate_gene_names_from_cfg( N_GENES, config_scene_df )
target_var = 'Y'
MCY_df.columns = var_names + [target_var]

Check correlations (**differences are indicated with 2 asterisks**):
check_MCY_correlations(MCY_df.copy())
``````
=============== FINAL CORRELATIONS ===============

==== 1. Group mean |correlation| of genes with Y_bin ====
======== Target: **A≈0.2  B≈0.5**  D≈0  NS≈0 ============
  A(signal)       **avg|r|=0.0210**  min|r|=0.0002  max|r|=0.0592
  B(signal)       **avg|r|=0.0151**  min|r|=0.0000  max|r|=0.0639
  D(weak)         avg|r|=0.0307  min|r|=0.0001  max|r|=0.0871
  NS(noise)       avg|r|=0.0171  min|r|=0.0001  max|r|=0.0597

==== 2. Spot-check: individual genes & covariates vs Y_bin ====
===== Target: A≈0.2  B≈0.5  age≈0.3  sex≈0.1  INR≈0  batch≈0 =====
  Variable                Cor(Y_bin)    Expected
  ------------------------------------------------
  A_gene1                    -0.0427        ~0.2
  A_gene2                    -0.0257        ~0.2
  B_gene102                   0.0244        ~0.5
  B_gene103                  -0.0122        ~0.5
  D_gene302                  -0.0177        ~0.0
  NS_gene602                 -0.0049        ~0.0
  NS_gene1000                 0.0221        ~0.0
  ------------------------------------------------
  age                        **-0.0265        ~0.3**
  sex                        **-0.0310        ~0.1**
  INR                        -0.0124        ~0.0
  batch                       0.0054        ~0.0
  ------------------------------------------------

==== 3. Average inter-gene |r| within groups ====
===== Target: A:0.3-0.6  B:0.2-0.4  D:0.2-0.3  NS:~0 =====
  A(signal)       avg|r|=0.3920  min|r|=0.3092  max|r|=0.4636   **Aceptable (desirable max|r|~0.6)**
  B(signal)       avg|r|=0.3745  min|r|=0.2893  max|r|=0.4605   **OK**
  D(weak)         avg|r|=0.2615  min|r|=0.1656  max|r|=0.3669   **OK**
  NS(noise)       avg|r|=0.0179  min|r|=0.0000  max|r|=0.1079   **OK**

==== 4. Covariate vs gene-group correlations ====
===== Target: age vs A≈0.4;  sex vs B≈0.4 =====
  Covariate-Group             avg|r|    min|r|    max|r|
  ------------------------------------------------------
  age vs A(signal)            0.3681    0.3258    0.4177  **OK**
  sex vs B(signal)            0.0139    0.0001    0.0514  **KO: ~0.4**
  ------------------------------------------------------ 
``````

**Screenshots**

**Desktop (please complete the following information):**
 - OS: [Windows 10]
 - Browser [firefox]
 - Version [calm_data_generator:  2.2.1]

**Smartphone (please complete the following information):**
 - Device: N/A
 - Version [calm_data_generator:  2.2.1]

**Additional context**

I used this function to check the correlations on final dataset:         
 def check_MCY_correlations(MCY_df: pd.DataFrame,
                           n_genes: int = 1000,
                           groups: dict = None) -> None:
    """
    Validates that the simulated MCY matrix has the expected correlation structure.

    Checks:
      1. Mean |r| of each gene group with Y_bin   (A≈0.2, B≈0.5, D≈0, NS≈0)
      2. Spot-check individual genes + all covariates vs Y_bin
      3. Average inter-gene |r| within each group  (A:0.3-0.6, B:0.2-0.4, D:0.2-0.3)
      4. Covariate vs gene-group |r|               (age vs A≈0.4, sex vs B≈0.4)

    Parameters
    ----------
    MCY_df  : DataFrame with gene columns first, then covariates, last column = 'Y'
    n_genes : number of gene columns (default 1000)
    groups  : optional dict {group_name: np.array of absolute column indices}
    """

    print("\n=============== FINAL CORRELATIONS ===============")

    # Standardize only continuous numeric columns
    # Pearson r is scale-invariant, but we standardize for consistency.
    # Binary columns (Y, sex, batch) are excluded to avoid misleading z-scores.
    skip_std = {"Y", "sex", "batch"}
    cols_to_std = [c for c in MCY_df.columns if c not in skip_std]

    MCY_df = MCY_df.copy()
    MCY_df[cols_to_std] = (
        (MCY_df[cols_to_std] - MCY_df[cols_to_std].mean(axis=0))
        / MCY_df[cols_to_std].std(axis=0, ddof=1)
    )

    # Gene group index ranges — derived from n_genes, not hardcoded
    nA  = 100
    nB  = 200
    nD  = 300
    # nNS = n_genes - nA - nB - nD       # = 400 for n_genes=1000

    A_idx  = np.arange(0,        nA)
    B_idx  = np.arange(nA,       nA + nB)
    D_idx  = np.arange(nA + nB,  nA + nB + nD)
    NS_idx = np.arange(nA + nB + nD, n_genes)

    if groups is None:
        groups = {
            "A(signal)": A_idx,
            "B(signal)": B_idx,
            "D(weak)"  : D_idx,
            "NS(noise)": NS_idx,
        }
    # Gene matrix and response
    genesM = MCY_df.iloc[:, :n_genes].to_numpy()   # shape (n_samples, n_genes)
    Y_bin  = MCY_df["Y"].to_numpy()                # binary 0/1

    # 1. Group mean |r| with Y_bin  — vectorized
    print("\n==== 1. Group mean |correlation| of genes with Y_bin ====")
    print(  "======== Target: A≈0.2  B≈0.5  D≈0  NS≈0 ============")

    for group_name, idxs in groups.items():
        # Stack genes as rows → corrcoef returns (n_genes+1 × n_genes+1) matrix
        # First row/col = Y_bin; remaining = genes
        block  = np.vstack([Y_bin, genesM[:, idxs].T])   # (1+k, n_samples)
        rmat   = np.corrcoef(block)                        # (1+k, 1+k)
        r_with_Y = np.abs(rmat[0, 1:])                    # |r| of each gene with Y

        print(f"  {group_name:<16}"
              f"avg|r|={r_with_Y.mean():.4f}  "
              f"min|r|={r_with_Y.min():.4f}  "
              f"max|r|={r_with_Y.max():.4f}")

    # 2. Spot-check individual genes + ALL covariates vs Y_bin
    spot_abs = {
        "A_gene1"    : 0,
        "A_gene2"    : 1,
        "B_gene102"  : 102,
        "B_gene103"  : 103,
        "D_gene302"  : 302,
        "NS_gene602" : 602,
        "NS_gene1000": 999,
    }
    # All covariates: expected correlations with Y
    covariate_expected = {
        "age"  : "~0.3",
        "sex"  : "~0.1",
        "INR"  : "~0.0",   # confounders not linked to Y
        "batch": "~0.0",
    }

    print("\n==== 2. Spot-check: individual genes & covariates vs Y_bin ====")
    print(  "===== Target: A≈0.2  B≈0.5  age≈0.3  sex≈0.1  INR≈0  batch≈0 =====")
    print(f"  {'Variable':<22}{'Cor(Y_bin)':>12}  {'Expected':>10}")
    print("  " + "-" * 48)

    for name, col in spot_abs.items():
        r = np.corrcoef(Y_bin, genesM[:, col])[0, 1]
        group = name.split("_")[0]
        expected = {"A": "~0.2", "B": "~0.5", "D": "~0.0", "NS": "~0.0"}.get(group, "")
        print(f"  {name:<22}{r:>12.4f}  {expected:>10}")

    print("  " + "-" * 48)

    for var, expected in covariate_expected.items():
        if var not in MCY_df.columns:
            print(f"  {var:<22}{'[column not found]':>12}")
            continue
        # For binary covariates (sex, batch), use the original integer codes
        vec = MCY_df[var].to_numpy()
        r   = np.corrcoef(Y_bin, vec)[0, 1]
        print(f"  {var:<22}{r:>12.4f}  {expected:>10}")

    print("  " + "-" * 48)

    # 3. Average inter-gene |r| within each group
    #    Uses the full upper-triangle of the within-group correlation matrix.
    print("\n==== 3. Average inter-gene |r| within groups ====")
    print(  "===== Target: A:0.3-0.6  B:0.2-0.4  D:0.2-0.3  NS:~0 =====")

    for group_name, idxs in groups.items():
        sub      = genesM[:, idxs]            # (n_samples, k)
        corr_mat = np.corrcoef(sub.T)          # (k, k)
        tri      = corr_mat[np.triu_indices(len(idxs), k=1)]
        abs_tri  = np.abs(tri)
        print(f"  {group_name:<16}"
              f"avg|r|={abs_tri.mean():.4f}  "
              f"min|r|={abs_tri.min():.4f}  "
              f"max|r|={abs_tri.max():.4f}")

    # 4. Covariate vs gene-group correlations
    #    Uses groups dict so custom group definitions are respected.
    print("\n==== 4. Covariate vs gene-group correlations ====")
    print(  "===== Target: age vs A≈0.4;  sex vs B≈0.4 =====")
    print(f"  {'Covariate-Group':<24}{'avg|r|':>10}{'min|r|':>10}{'max|r|':>10}")
    print("  " + "-" * 54)

    covariate_group_pairs = [
        ("age", "A(signal)"),
        ("sex", "B(signal)"),
    ]
    for var, group_name in covariate_group_pairs:
        if var not in MCY_df.columns or group_name not in groups:
            print(f"  {var} vs {group_name:<20}  [skipped — column or group not found]")
            continue
        idxs      = groups[group_name]
        vec       = MCY_df[var].to_numpy()
        # Vectorized: stack covariate + gene block
        block     = np.vstack([vec, genesM[:, idxs].T])
        rmat      = np.corrcoef(block)
        r_vals    = np.abs(rmat[0, 1:])
        label     = f"{var} vs {group_name}"
        print(f"  {label:<24}"
              f"{r_vals.mean():>10.4f}"
              f"{r_vals.min():>10.4f}"
              f"{r_vals.max():>10.4f}")

    print("  " + "-" * 54)

**Comment by AlejandroBeldaFernandez** [comment]
Resolved. Try it and let me know. Thanks for the feedback.

**Comment by soyunalmapresente** [comment]
**Not Resolved**. I have not permission to reopen the issue.  
  
Hello Again. First of all thanks for the quick answer.
I reinstalled with          pip install --upgrade --no-cache-dir calm-data-generator
Regenerated the dataset
But the result is still the same.
``````
=============== FINAL CORRELATIONS ===============

==== 1. Group mean |correlation| of genes with Y_bin ====
======== Target: **A≈0.2  B≈0.5**  D≈0  NS≈0 ============
  A(signal)       **avg|r|=0.0210**  min|r|=0.0002  max|r|=0.0592 **KO**
  B(signal)       **avg|r|=0.0151**  min|r|=0.0000  max|r|=0.0639 **KO**
  D(weak)         avg|r|=0.0307  min|r|=0.0001  max|r|=0.0871
  NS(noise)       avg|r|=0.0171  min|r|=0.0001  max|r|=0.0597

==== 2. Spot-check: individual genes & covariates vs Y_bin ====
===== Target: A≈0.2  B≈0.5  age≈0.3  sex≈0.1  INR≈0  batch≈0 =====
  Variable                Cor(Y_bin)    Expected
  ------------------------------------------------
  A_gene1                    -0.0427        ~0.2
  A_gene2                    -0.0257        ~0.2
  B_gene102                   0.0244        ~0.5
  B_gene103                  -0.0122        ~0.5
  D_gene302                  -0.0177        ~0.0
  NS_gene602                 -0.0049        ~0.0
  NS_gene1000                 0.0221        ~0.0
  ------------------------------------------------
  age                        **-0.0265        KO ~0.3**
  sex                        **-0.0310        KO ~0.1**
  INR                        -0.0124        ~0.0
  batch                       0.0054        ~0.0
  ------------------------------------------------

==== 3. Average inter-gene |r| within groups ====
===== Target: A:0.3-0.6  B:0.2-0.4  D:0.2-0.3  NS:~0 =====
  A(signal)       avg|r|=0.3920  min|r|=0.3092  max|r|=0.4636   **Aceptable (desirable max|r|~0.6)**
  B(signal)       avg|r|=0.3745  min|r|=0.2893  max|r|=0.4605
  D(weak)         avg|r|=0.2615  min|r|=0.1656  max|r|=0.3669
  NS(noise)       avg|r|=0.0179  min|r|=0.0000  max|r|=0.1079

==== 4. Covariate vs gene-group correlations ====
===== Target: age vs A≈0.4;  sex vs B≈0.4 =====
  Covariate-Group             avg|r|    min|r|    max|r|
  ------------------------------------------------------
  age vs A(signal)            0.3681    0.3258    0.4177
  sex vs B(signal)            0.0139    0.0001    0.0514  **KO: ~0.4**
``````
